core/internal/filetransfer/file_transfer_gcs.go

```diff
@@ -25,34 +37,291 @@ type GCSFileTransfer struct {
+
+	// open the file for writing and defer closing it
+	file, err := os.Create(localPath)
+	if err != nil {
+		return formatDownloadError(
+			fmt.Sprintf("error creating download file %s", localPath),
+			err,
+		)
+	}
+
+	defer func(file *os.File) {
+		fileError := file.Close()
+		if err == nil {
+			err = fileError
+		}
+	}(file)
+
+	_, err = io.Copy(file, r)
+	if err != nil {
+		return formatDownloadError(
+			fmt.Sprintf("error copying file %s", localPath),
+			err,
+		)
+	}
+
+	return nil
+}
+
+// getObjectAndAttrs returns the object handle and attributes if the object
+// exists and corresponds with a file.
+//
+// If the object corresponds to a directory, we return the ErrObjectIsDirectory
+// error to skip downloading without failing.
+func (ft *GCSFileTransfer) getObjectAndAttrs(
+	bucket *storage.BucketHandle,
+	task *ReferenceArtifactDownloadTask,
+	key string,
+) (*storage.ObjectHandle, *storage.ObjectAttrs, error) {
+	object, err := ft.getObject(bucket, task, key)
+	if err != nil {
+		return nil, nil, err
+	}
+	if strings.HasSuffix(object.ObjectName(), "/") {
+		return nil, nil, ErrObjectIsDirectory
+	}
+
+	attrs, err := object.Attrs(ft.ctx)
+	// if object doesn't have an extension and has size 0, it might be a
+	// directory as we cut off the ending "/" in the manifest entry.
+	if err != nil && (path.Ext(object.ObjectName()) == "" && task.Size == 0) {
+		key := key + "/"
+		object, err := ft.getObject(bucket, task, key)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// if the new key corresponds with an existing directory, it should
+		// return the attrs without an error
+		_, err = object.Attrs(ft.ctx)
+		if err != nil {
+			return nil, nil, err
+		}
+		return nil, nil, ErrObjectIsDirectory
+	} else if err != nil {
+		return nil, nil, err
+	}
+	return object, attrs, nil
+}
+
+// getObject returns the object handle given a bucket, key, and task object
+// with a possible versionId.
+func (ft *GCSFileTransfer) getObject(
+	bucket *storage.BucketHandle,
+	task *ReferenceArtifactDownloadTask,
+	key string,
+) (*storage.ObjectHandle, error) {
+	object := bucket.Object(key)
+	if task.VersionId != nil {
+		versionId, err := safeConvertToInt64(task.VersionId)
+		if err != nil {
+			err := fmt.Errorf("failed to convert VersionId: %v", err)
+			return nil, err
+		}
+		object = object.Generation(versionId)
+	}
+	return object, nil
+}
+
+// parseReference parses the reference path and returns the bucket name and
+// object name.
+func parseReference(reference string) (string, string, error) {
+	uriParts, err := url.Parse(reference)
+	if err != nil {
+		return "", "", err
+	}
+	if uriParts.Scheme != "gs" {
+		err := fmt.Errorf("invalid gsutil URI %s", reference)
+		return "", "", err
+	}
+	bucketName := uriParts.Host
+	objectName := strings.TrimPrefix(uriParts.Path, "/")
+	return bucketName, objectName, nil
+}
+
+// getDownloadFilePath returns the file path to download the file to when
+// removing duplicate info from the extension.
```

Comment:
ah sorry this is pretty confusing, I'll word it better!

but basically if the reference path is `gs://path/to/folder` with `gs://path/to/folder/file.txt` in it, the download path we pass will be something like `artifacts/ref-art:v0/folder` which was formed based on the reference path. so the function would remove the substring prefix `gs://path/to/folder` from `gs://path/to/folder/file.txt` and get `/file.txt`, which then it would add to the original path.